packages/flutter_tools/lib/src/ios/mac.dart

```diff
@@ -742,15 +707,48 @@ void _handleXCResultIssue({required XCResultIssue issue, required Logger logger}
       break;
   }
 
-  // Add more custom output for flutter users.
-  if (issue.message != null && issue.message!.toLowerCase().contains('provisioning profile')) {
-    logger.printError('');
-    logger.printError('It appears that there was a problem signing your application prior to installation on the device.');
-    logger.printError('');
-    logger.printError('Verify that the Bundle Identifier in your project is your signing id in Xcode');
-    logger.printError('  open ios/Runner.xcworkspace');
-    logger.printError('');
-    logger.printError("Also try selecting 'Product > Build' to fix the problem:");
+  // Add more error messages for flutter users for some special errors.
+  if (issue.message != null ) {
+    final String message = issue.message!;
+    if (message.toLowerCase().contains('requires a provisioning profile.')) {
+      logger.printError(noProvisioningProfileInstruction, emphasis: true);
+    } else if(message.toLowerCase().contains('provisioning profile')) {
+      logger.printError(noDevelopmentTeamInstruction, emphasis: true);
+    }
+  }
+}
+
+// Returns `true` if at least one issue is detected. `false` if xcresult parsing was failed or no issues are detected.
+bool _handleIssuesFromXCResult(XCResult xcResult, Logger logger) {
+  if (xcResult.parseSuccess) {
+    bool issueDetected = false;
+    for (final XCResultIssue issue in xcResult.issues) {
+      _handleXCResultIssue(issue: issue, logger: logger);
+      issueDetected = true;
+    }
+    return issueDetected;
+  } else {
+    globals.printTrace('XCResult parsing error: ${xcResult.parsingErrorMessage}');
+    return false;
+  }
+}
+
+// Detects and handles errors without using xcresult.
+void  _handleIssuesWithoutUsingXCResult(XcodeBuildExecution xcodeBuildExecution, Logger logger, XcodeBuildResult result) {
+  if (xcodeBuildExecution.environmentType == EnvironmentType.physical
+      // May need updating if Xcode changes its outputs.
+      && (result.stdout?.contains('requires a provisioning profile. Select a provisioning profile in the Signing & Capabilities editor') ?? false)) {
+    logger.printError(noProvisioningProfileInstruction, emphasis: true);
+    return;
+  }
+  // Make sure the user has specified one of:
+  // * DEVELOPMENT_TEAM (automatic signing)
+  // * PROVISIONING_PROFILE (manual signing)
+  if (xcodeBuildExecution.environmentType == EnvironmentType.physical &&
+      !<String>['DEVELOPMENT_TEAM', 'PROVISIONING_PROFILE'].any(
```

Comment:
This seems like a better situation to use `noDevelopmentTeamInstruction` instead of `message.toLowerCase().contains('provisioning profile')` since it actually checks if you have `DEVELOPMENT_TEAM` or `PROVISIONING_PROFILE` set.  Can we pipe the `xcodeBuildExecution` into the xcresult handler to do similar checking?

Reply:
This is technically not part of handing xcresult. XCResult does detect this error by showing a message: ```
Showing Recent Messages
Signing for "Runner" requires a development team. Select a development team in the Signing & Capabilities editor.

```

How about move this part where we directly check the flags all the way above xcresult handling? 

So the flow becomes:
1. Check `DEVELOPMENT_TEAM` and `PROVISIONING_PROFILE` flags.
2. Handle xcresults
3. Handle other issues with stdout.

Reply:
> I'm not following this, why would this one example be a special case where it's handled before the message is parsed?

So the "Check DEVELOPMENT_TEAM and PROVISIONING_PROFILE flag" can be done regardless if xcresult's parsing is successful or not. If this check is part of xcresult handling, it will not be run when xcresult is failed to parse. Or we do the same thing in xcresult handling as well as the fallback error handling when xcresult is failed to parse?

Reply:
I'm not sure how you want to organize it 🙂  We should only print out the same error once, and include useful recovery suggestions where possible.  There's currently fallback to string parsing if the result bundle fails to build, but maybe that never happens (or we should add `Usage` data about it to see).